drivers/can/can_numaker.c

```diff
@@ -0,0 +1,316 @@
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(can_numaker, CONFIG_CAN_LOG_LEVEL);
+
+/* Implementation notes
+ * 1. Use Bosch M_CAN driver (m_can) as backend
+ *    NOTE: m_can only supports single configuration of max element numbers for all instances
+ * 2. Need to modify can_numaker_get_core_clock() for new SOC support
+ */
+
+struct can_numaker_config {
+	mm_reg_t   canfd_base;
+	mem_addr_t  mrba;
+	mem_addr_t  mram;
+	const struct reset_dt_spec reset;
+	uint32_t    clk_modidx;
+	uint32_t    clk_src;
+	uint32_t    clk_div;
+	const struct device *clk_dev;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+};
+
+struct can_numaker_data {
+	/* Avoid zero length struct */
+	uint32_t resv0;
+};
+
+static int can_numaker_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc = 0;
+	uint32_t clksrc_rate_idx;
+	uint32_t clksrc_rate;
+	uint32_t clkdiv_divider;
+
+	/* Initialize clock source rate table
+	 * NOTE: Fix the table for new SOC support
+	 */
+	uint32_t clksrc_rate_tab[] = {
+		__HXT,      /* HXT */
+		0,          /* PLLFOUT/2 */
+		0,          /* HCLK */
+		__HIRC      /* HIRC */
+	};
+	clksrc_rate_tab[1] = CLK_GetPLLClockFreq() / 2;
+	clksrc_rate_tab[2] = CLK_GetHCLKFreq();
+
+	/* SOC in support list
+	 * NOTE: * NOTE: Fix the conditional for new SOC support
+	 */
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_M46X)) {
+		LOG_ERR("Not support SOC: " CONFIG_SOC_SERIES);
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Module clock source rate */
+	clksrc_rate_idx = CLK_GetModuleClockSource(config->clk_modidx);
+	if (clksrc_rate_idx >= ARRAY_SIZE(clksrc_rate_tab)) {
+		LOG_ERR("Invalid clock source rate table or index");
+		rc = -EIO;
+		goto cleanup;
+	}
+	clksrc_rate = clksrc_rate_tab[clksrc_rate_idx];
+
+	/* Module clock divider */
+	clkdiv_divider = CLK_GetModuleClockDivider(config->clk_modidx) + 1;
+
+	LOG_DBG("Clock source/divider: %d/%d", clksrc_rate, clkdiv_divider);
+	*rate = clksrc_rate / clkdiv_divider;
+
+cleanup:
+
+	return rc;
+}
+
+static int can_numaker_init(const struct device *dev)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	struct numaker_scc_subsys scc_subsys;
+	int rc = 0;
+	uint32_t rate;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id        = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx   = config->clk_modidx;
+	scc_subsys.pcc.clk_src      = config->clk_src;
+	scc_subsys.pcc.clk_div      = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
```

Comment:
Why are these comments here? These are Zephyr APIs. No need to mention what they would be equivalent to in some other API.

Reply:
The purpose is to record the migration from proprietary CLK control to standard one.
I see and not need mention the proprietary CLK API.